src/test/lib/concurrency/stress_test.cpp

```diff
@@ -695,4 +710,104 @@ TEST_F(StressTest, VisibilityOfInsertsBeingRolledBack) {
+  // Create a table with multiple TableKeyConstraints.
+  auto table = std::make_shared<Table>(
+      TableColumnDefinitions{{"a", DataType::Int, false}, {"b", DataType::Int, false}, {"c", DataType::Int, false}},
+      TableType::Data, std::nullopt, UseMvcc::Yes);
+  primary_key_constraint(table, {"a"});
+
+  table->append({1, 1, 1});
+  table->append({2, 2, 2});
+  table->append({3, 3, 1});
+
+  Hyrise::get().storage_manager.add_table("dummy_table", table);
+  /** This test runs insertions and reads concurrently. Specifically, it tests the following functions:
+   * - `UccDiscoveryPlugin::_validate_ucc_candidates`
+   * - `StoredTableNode::unique_column_combinations`
+   * In order to simulate insertions parallel to the reads, we have to clear the constraints in the table. As this is
+   * only needed for the test, we can use a mutex to ensure that this does not happen in parallel.
+   */
+
+  Hyrise::get().default_pqp_cache = std::make_shared<SQLPhysicalPlanCache>();
+  Hyrise::get().default_lqp_cache = std::make_shared<SQLLogicalPlanCache>();
+  auto& pm = Hyrise::get().plugin_manager;
+
+  pm.load_plugin(build_dylib_path("libhyriseUccDiscoveryPlugin"));
+
+  auto deletion_mutext = std::shared_mutex{};
+
+  auto start_flag = std::atomic_flag{};
+  auto stop_flag = std::atomic_flag{};
+
+  const auto VALIDATION_COUNT = uint32_t{100};
+  const auto SLEEP_TIME = std::chrono::milliseconds{1};
+  auto writer_waiting = std::atomic<bool>{false};
+
+  const auto validate_constraint = [&] {
+    start_flag.wait(false);
+    for (auto i = uint32_t{0}; i < VALIDATION_COUNT; ++i) {
+      // Populate the plan cache.
+      const auto sql = std::string{"SELECT b,c FROM dummy_table GROUP BY b,c;"};
+      auto pipeline = SQLPipelineBuilder{sql}.create_pipeline();
+      pipeline.get_result_table();
+
+      pm.exec_user_function("hyriseUccDiscoveryPlugin", "DiscoverUCCs");
+
+      std::this_thread::sleep_for(SLEEP_TIME);
+
+      // Notify the reader threads that the writer is waiting.
+      writer_waiting = true;
+      const auto lock = std::unique_lock{deletion_mutext};
+      clear_soft_key_constraints(table);
+      writer_waiting = false;
+    }
+  };
+
+  const auto stored_table_node_constraint_access = [&] {
+    start_flag.wait(false);
+    while (!stop_flag.test()) {
+      const auto stored_table_node = std::make_shared<StoredTableNode>("dummy_table");
+      // Access the unique column combinations. We need to lock here because `unique_column_combinations` uses a
+      // reference to iterate over the constraints. This reference is invalidated when the constraints are cleared.
+      while (writer_waiting) {
+        std::this_thread::sleep_for(SLEEP_TIME);
+      }
+      const auto lock = std::shared_lock{deletion_mutext};
+      ASSERT_LE(stored_table_node->unique_column_combinations().size(), 3);
+    }
+  };
+
+  // Start running the different modifications in parallel.
+  const auto thread_count = 10;
```

Comment:
Guess you can easily do 50 or 100 here. Just double-check the (sanitizer) runtime in the pipeline.